reflection/serverreflection.go

```diff
@@ -65,291 +60,165 @@ type GRPCServer interface {
 	GetServiceInfo() map[string]grpc.ServiceInfo
 }
 
+// ExtensionResolver is the interface used to query details about extensions.
+// This interface is satisfied by protoregistry.GlobalTypes.
+type ExtensionResolver interface {
+	protoregistry.ExtensionTypeResolver
+	RangeExtensionsByMessage(message protoreflect.FullName, f func(protoreflect.ExtensionType) bool)
+}
+
 var _ GRPCServer = (*grpc.Server)(nil)
 
 type serverReflectionServer struct {
 	rpb.UnimplementedServerReflectionServer
-	s GRPCServer
+	s            GRPCServer
+	descResolver protodesc.Resolver
+	extResolver  ExtensionResolver
 
-	initSymbols  sync.Once
-	serviceNames []string
-	symbols      map[string]*dpb.FileDescriptorProto // map of fully-qualified names to files
+	initServiceNames sync.Once
+	serviceNames     []string
 }
 
-// Register registers the server reflection service on the given gRPC server.
-func Register(s GRPCServer) {
-	rpb.RegisterServerReflectionServer(s, &serverReflectionServer{
-		s: s,
-	})
+// ServerOptions represents the options used to construct a reflection server.
+//
+// Either Server or ServiceNames must be populated, but not both. These control
+// what services are advertised by the server in the ListServices capability of
+// the reflection service. If neither is provided, the returned server can still
+// serve descriptors, but it will advertise no service names.
+//
+// The given DescriptorResolver will be used to resolve symbols and files by
+// name. If not present, protoregistry.GlobalFiles will be used. The given
+// ExtensionResolver will be used to resolve extensions. If not present,
+// protoregistry.GlobalTypes will be used.
+type ServerOptions struct {
+	// An RPC server, whose exposed services are made available via service
+	// reflection.
+	Server GRPCServer
+	// The list of service names. This should only be populated if Server is
+	// nil.
+	ServiceNames []string
+	// Optional resolver used to load descriptors.
+	DescriptorResolver protodesc.Resolver
+	// Optional resolver used to query for known extensions.
+	ExtensionResolver ExtensionResolver
```

Comment:
Let's document the default value for these. `If unset, protoregistry.GlobalFiles will be used`

Reply:
These were documented above. So I rearranged a little and moved the relevant comments down to the fields' Go doc.